DNN Platform/Library/Startup.cs

```diff
@@ -30,8 +30,14 @@ public class Startup : IDnnStartup
         public void ConfigureServices(IServiceCollection services)
         {
             services.AddSingleton<WebFormsModuleControlFactory>();
+            services.AddSingleton<IModuleControlFactory>(p => p.GetRequiredService<WebFormsModuleControlFactory>());

```

Comment:
Is this registration correct?  Since this registers a singleton with the interface does this result in proper behavior?  

As belowon on lines 36 and 39 the same type is being injected with a different implementation

Reply:
Yes, this is correct.

In the library `Microsoft.Extensions.DependenyInjection` it's possible to register the same type multiple times. For example:

```cs
interface IService {}
class ServiceA : IService {}
class ServiceB : IService {}

IServiceCollection services;
services.AddSingleton<IService, ServiceA>();
services.AddSingleton<IService, ServiceB>();
```

Whenever you resolve the service `IService`, you'll get the last registered service. So the following would resolve ServiceB:
```cs
IServiceProvider provider;
provider.GetRequiredService<IService>(); // Resolved ServiceB
```

However, it's also possible to get all the registered services of type `IService` with `IEnumerable<IService>`:
```cs
IServiceProvider provider;
provider.GetRequiredService<IEnumerable<IService>>(); // An enumerable with ServiceA and ServiceB
```

This means, if an extension registers their custom type:
```cs
class MyService : IService {}

IServiceCollection services;
services.AddSingleton<IService, MyService>();
```
It'll add the MyService to the IEnumerable.

This is what I'm doing in ModuleControlPipeline:
https://github.com/dnnsoftware/Dnn.Platform/blob/05852bfe270203f4fe4a0e291ce560e6e96e3f70/DNN%20Platform/DotNetNuke.ModulePipeline/ModuleControlPipeline.cs#L41-L46

Hence why I'm registering WebFormsModuleControlFactory, Html5ModuleControlFactory and ReflectedModuleControlFactory this way.

---

For the reason why I'm not doing the following:

```cs
services.AddSingleton<WebFormsModuleControlFactory>();
services.AddSingleton<IModuleControlFactory, WebFormsModuleControlFactory>();
```

When you resolve `WebFormsModuleControlFactory` it'll create a new instance and when you resolve `IEnumerable<IModuleControlFactory>`, it'll create another instance. The dependency injector doesn't try to check if there is already an instance of the registered type.

By doing the following:

```cs
services.AddSingleton<WebFormsModuleControlFactory>();
services.AddSingleton<IModuleControlFactory>(p => p.GetRequiredService<WebFormsModuleControlFactory>());
```

It'll create an instance if you resolve `WebFormsModuleControlFactory` and when you try to resolve `IEnumerable<IModuleControlFactory>`, it'll get the instance of the `WebFormsModuleControlFactory` from the service provider. This way you won't get multiple instances of the factory in memory.

If you want to read more, this is better documented by Microsoft than explained by me: https://learn.microsoft.com/en-us/dotnet/core/extensions/dependency-injection#service-registration-methods

Reply:
I've been using [`TryAddEnumerable`](https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.dependencyinjection.extensions.servicecollectiondescriptorextensions.tryaddenumerable?view=dotnet-plat-ext-7.0&viewFallbackFrom=netframework-4.7.2) when I'm registering multiple implementations, just to make it clearer that it's intentional.

I wasn't aware about the issue with getting multiple instances of singletons, that's good to know.

Reply:
> Since we're no longer injecting these into `ModuleControlPipeline`, do we need the standalone `AddSingleton<WebFormsModuleControlFactory>()`?

We can remove them, yes.

The only issue is when external extensions are resolving `WebFormsModuleControlFactory` in their code, other than that we can remove them. But I highly doubt it that anyone would use the factories directly.

Reply:
> I wasn't aware about the issue with getting multiple instances of singletons, that's good to know.

Sadly using the `GetRequiredService`-workaround also brings a problem with disposing the class.

For example the following (Fiddle: https://dotnetfiddle.net/2UVbpb)
```cs
using System;
using Microsoft.Extensions.DependencyInjection;

var services = new ServiceCollection();
services.AddSingleton<Service>();
services.AddSingleton<IService>(p => p.GetRequiredService<Service>());

var provider = services.BuildServiceProvider();
provider.GetRequiredService<IService>();
provider.Dispose();

interface IService {}

class Service : IService, IDisposable
{
	public Service() => Console.WriteLine("Construct");
	public void Dispose() => Console.WriteLine("Dispose");
}
```

Would dispose the class twice. There is an open issue about this problem: https://github.com/dotnet/runtime/issues/70004 but sadly it is not addressed yet.